interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java

```diff
@@ -0,0 +1,1937 @@
+            } else if (isInt64(inputType)) {
+                return box(unboxLong(input), outputType);
+            }
+        } else if (isUnsigned(outputType)) {
+            if (isInt8(outputType)) {
+                return box(unboxInt(input) & 0xff, outputType);
+            } else if (isInt16(inputType)) {
+                return box(unboxInt(input) & 0xffff, outputType);
+            } else if (isInt32(inputType)) {
+                return box(unboxInt(input) & 0xffff_ffffL, outputType);
+            } else if (isInt64(inputType)) {
+                return box(unboxLong(input), outputType);
+            }
+        } else if (isFloat64(inputType) && isFloat32(outputType)) {
+            return box((float) unboxDouble(input), outputType);
+        }
+        throw new IllegalStateException("Invalid truncate");
+    }
+
+    @Override
+    public Object visit(VmThreadImpl param, Xor node) {
+        Value left = node.getLeftInput();
+        Value right = node.getRightInput();
+        ValueType inputType = left.getType();
+        assertSameTypes(node);
+        if (isInt64(inputType)) {
+            return box(unboxLong(left) ^ unboxLong(right), node.getType());
+        } else if (isInt32(inputType)) {
+            return box(unboxInt(left) ^ unboxInt(right), node.getType());
+        } else if (isBool(inputType)) {
+            return Boolean.valueOf(unboxBool(left) ^ unboxBool(right));
+        }
+        throw badInputType();
+    }
+
+    //
+
+    ////////////
+    // Literals
+    ////////////
+
+    @Override
+    public Object visit(VmThreadImpl param, ArrayLiteral node) {
+        MemoryImpl memory = new MemoryImpl((int) node.getType().getSize());
+        List<Literal> nodeValues = node.getValues();
+        ValueType elementType = node.getType().getElementType();
+        long elementSize = node.getType().getElementSize();
+        for (int i = 0; i < nodeValues.size(); i++) {
+            Literal value = nodeValues.get(i);
+            store(memory, (int) (elementSize * i), elementType, value, MemoryAtomicityMode.NONE);
+        }
+        return memory;
+    }
+
+    @Override
+    public Object visit(VmThreadImpl param, BooleanLiteral node) {
+        return Boolean.valueOf(node.booleanValue());
+    }
+
+    @Override
+    public Object visit(VmThreadImpl param, FloatLiteral node) {
+        return box(node.doubleValue(), node.getType());
+    }
+
+    @Override
+    public Object visit(VmThreadImpl param, IntegerLiteral node) {
+        return box(node.longValue(), node.getType());
+    }
+
+    @Override
+    public Object visit(VmThreadImpl param, ObjectLiteral node) {
+        return node.getValue();
+    }
+
+    @Override
+    public Object visit(VmThreadImpl thread, StringLiteral node) {
+        VmClassLoaderImpl classLoader = (VmClassLoaderImpl) element.getEnclosingType().getContext().getClassLoader();
+        return classLoader.getString(thread, node.getValue(), true);
```

Comment:
Does this globally intern the string?  All string constants need to be interned but we can add that as a follow on task as well

Reply:
The BuildTimeHeap code will intern StringLiterals when it emits them into the native image.  Maybe that is good enough (if we define compile time `equals` between StringLiteral values appropriately)?

Reply:
User code can trip up on this though.... with the interpreter not recognizing two strings as `==` that should be, or with the final native image treating two strings as `==` when it shouldn't.  It's annoying, but I think we need to follow the programmer's intent on this.

The BuildTimeHeap can intern common `char[]'s from strings with the same content but different identity, but it probably shouldn't remove the identity differences.

Reply:
In BuildTimeHeap, I'm interring StringLiteral but not interning instances of java.lang.String that come through as ObjectLiterals.  My assumption was that a StringLiteral would only be coming from a string literal in a class file, and therefore should be interned by the system (while ObjectLiteral would be interned by the program).

Reply:
You're right about BuildTimeHeap being OK to intern `.equals` `StringLiterals`.  The only code I can find that creates `StringLiterals` is in ClassFileImpl and only for constantpool entries.

The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image

Reply:
Yeah only string literals are intentionally interned, and certain other specific cases like class names and so forth.  I do agree that string literals should be interned on a per-class-loader basis.

Reply:
> The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image

Sorry if I wasn't clear - this interpreter behaviour is a bug.  per-class-loader interning of strings is contrary to the spec

Reply:
> > The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image
> 
> Sorry if I wasn't clear - this interpreter behaviour is a bug. per-class-loader interning of strings is contrary to the spec

Ah okay, well good to know.  I'll get that fixed...